### ISSUE TITLE
i3status/blocks bar hangs up from time to time

### ISSUE THREAD

**Issue by unera** [i3/i3#3242]
From time to time i3status/blocks bar hangs up.
I think the bar hangs often after suspend/resume. If CPU usage reaches 100% the bar hangs up, too.

Example: Now is 10:57, but the bar did hang at 10:39:

![](https://hsto.org/webt/ac/la/d9/aclad97r6yzuk0rl8cbmi-7rg8c.png)

PS:
Binary i3 version:  4.13 (2016-11-08) © 2009 Michael Stapelberg and contributors
Running i3 version: 4.13 (2016-11-08) (pid 5895)o abort…)
Loaded i3 config: /home/unera/.i3/config (Last modified: Ср 07 мар 2018 21:56:48, 3329319 seconds ago)

The i3 binary you just called: /usr/bin/i3
The i3 binary you are running: i3

https://logs.i3wm.org/logs/5703935746637824.bz2

**Comment by i3bot** [comment]
Sorry, we can only support the latest major version. Please upgrade from 4.13 to 4.15, verify the bug still exists, and re-open this issue.

**Comment by unera** [comment]
Please reopen the ticket. 4.15 version hangs up, too:

logfile: https://logs.i3wm.org/logs/5660082654150656.bz2

$ i3 --moreversion 2>&- || i3 --version                                          
Binary i3 version:  4.15 (2018-03-10) © 2009 Michael Stapelberg and contributors
Running i3 version: 4.15 (2018-03-10) (pid 5895)o abort…)
Loaded i3 config: /home/unera/.i3/config (Last modified: Ср 07 мар 2018 21:56:48, 3330659 seconds ago)

The i3 binary you just called: /usr/bin/i3
The i3 binary you are running: i3

![](https://habrastorage.org/webt/2l/zs/hj/2lzshjwq0xexjfmkmbtxpsa6slg.png)

**Comment by unera** [comment]
```txt
$ cat .i3/blocks
# i3blocks config file
#
# Please see man i3blocks for a complete reference!
# The man page is also hosted at http://vivien.github.io/i3blocks
#
# List of valid properties:
#
# align
# color
# command
# full_text
# instance
# interval
# label
# min_width
# name
# separator
# separator_block_width
# short_text
# signal
# urgent

# Global properties
separator_block_width=15

[ticket]
command=test -r ~/.ticket && cat ~/.ticket
interval=1
color=#7777FF

[wifi]
instance=wlan0
interval=2
command=/usr/share/i3blocks/wifi
label=WiFi
min_width='100%'
separator=false

[wifi]
instance=wlp1s0
interval=2
command=/usr/share/i3blocks/wifi
label=W
min_width='100%'
separator=false

[wifi-name]
interval=2
command=sudo wpa_cli status|grep -i ^ssid=|sed 's/^ssid=//i'

[iface]
instance=wlan0
interval=2
command=/usr/share/i3blocks/iface
label=IP

[volume]
label=♪
instance=Master
command=/usr/share/i3blocks/volume
interval=1
signal=10

[temperature]
label=t
interval=10
command=/usr/share/i3blocks/temperature
separator=false

[cpu]
label=CPU
command=/usr/share/i3blocks/cpu_usage -w 30 -c 70
interval=2
separator=false
min_width=' 100.00% '
align=right

[battery]
label=⚡
command=/usr/share/i3blocks/battery
interval=30

[time]
command=date +'   %A, %d %b %H:%M:%S  '
interval=1
color=#8888ff

[lock]
command=bash ~/.scripts/lockscreen-bar
interval=once
color=#ee7777

[LANG]
command=~/.scripts/i3block-layout
interval=1
```

**Comment by unera** [comment]
> You said it also happens with i3status

time ago I opened the same ticket https://github.com/i3/i3/issues/1904 (i3 4.10)

I used i3status then. 

The ticket was fixed and I used 4.10. 
Last week I upgraded i3 to 4.13 (Debian/stable), then to 4.15.

statusbar hangs up again.

The bar often hangs after suspend/resume, but from time to time hangs if CPU usage reaches 100%.
Alt-Shift-r (i3 restart) solve the problem.

PS: I switched my i3 to i3status, https://logs.i3wm.org/logs/5731644862365696.bz2

the bar hangs, too:

![](https://habrastorage.org/webt/e5/7h/ul/e57hultemsc8_qfp7znwchyabjm.png)

my i3status config:
```txt
# i3status configuration file.
# see "man i3status" for documentation.

# It is important that this file is edited as UTF-8.
# The following line should contain a sharp s:
# ß
# If the above line is not correctly displayed, fix your editor first!

general {
        colors = true
        interval = 1
        output_format = "i3bar"
}

order += "wireless wlan0"
order += "wireless wlan1"
order += "run_watch DHCP"
order += "battery 0"
# order += "load"
order += "cpu_usage"
order += "volume_master"
order += "tztime local"

cpu_usage {
	format = "%usage"
}

wireless wlan0 {
        format_up = "W: (%quality at %essid) %ip"
        format_down = ""
}

wireless wlan1 {
        format_up = "W: (%quality at %essid) %ip"
        format_down = ""
}

battery 0 {
        format = "%status %percentage %remaining"
}

run_watch DHCP {
        pidfile = "/var/run/dhclient*.pid"
}

tztime local {
        format = "  %A, %d %b %H:%M:%S   "
}

load {
        format = "%1min"
}

disk "/" {
        format = "%avail"
}

volume master {
	format = "♪ %volume"
	format_muted = "♪ muted (%volume)"
	device = "default"
	mixer = "Master"
	mixer_idx = 0
}
volume master {
	format = "♪ %volume"
	format_muted = "♪ muted (%volume)"
	device = "default"
	mixer = "PCM"
	mixer_idx = 0
}
```

**Comment by Airblader** [comment]
Thanks for the confirmation that you can reproduce it with i3status, too.

**Comment by orestisf1993** [comment]
Does your i3bar update correctly? ie, do the workspace buttons update?

**Comment by unera** [comment]
> Does your i3bar update correctly? ie, do the workspace buttons update?


While i3bar hangs, I can switch (by mouse) through workspace, I can watch systray icons (ex. telegram icon on screenshot).

Only bar (i3block/i3status) hangs until Alt-Shift-r (i3 restart)ю

**Comment by orestisf1993** [comment]
Can you run i3status from the command line and see if you can get it to freeze?

**Comment by unera** [comment]
Is there any chanсe the somebody will pay attention with my ticket?

![ss](https://user-images.githubusercontent.com/538980/40265519-713e7724-5b42-11e8-87f3-b466ade5f496.png)

**Comment by orestisf1993** [comment]
I did pay attention to your ticket and regularly check my bar for hangs. Unfortunately, these issues are not easy to reproduce. If you have any idea why this is happening you are more than welcome to write a patch.

**Comment by unera** [comment]
> I did pay attention to your ticket and regularly check my bar for hangs. Unfortunately, these issues are not easy to reproduce.

The problem is well reproducible on 3 (three) my computers.

I use:
* Debian/stable,
* sysvint (not systemd),
* uswsusp,
* laptop-mode-tools
* acpi

each pm-suspend start has ~30% chance to see the bug.

**Comment by unera** [comment]
Also version 4.10 never hangs.

**Comment by unera** [comment]
Additional information:

Now my i3block has hung up.

I looked through `ps ax`, found the situation:

```txt

11938 ?        T      0:00 /bin/sh -c i3blocks -c ~/.i3/blocks
11939 ?        T      0:02 i3blocks -c /home/unera/.i3/blocks

```

then I did:

```txt
kill -CONT 11939
```

And my i3block began work without i3 restart.

So: i3 hangs because it received SIGSTOP, but SIGCONT was lost somewhere.

**Comment by unera** [comment]
Hi!

I've just installed Debian/i3 to new laptop and the bug is reproducible, too.

**Comment by stapelberg** [comment]
SIGSTOP/SIGCONT are only sent when i3bar goes into hide mode (so that the child process doesn’t unnecessarily compute updates which are invisible anyway).

Are you using i3bar’s hide mode? If yes, can you try going into hide mode and back to bar mode when you have the problem and see if that gets the process unstuck? If no, can you run `strace -f -p $(pidof i3bar) -y -o /tmp/st -s 2048` and toggle hide mode/bar mode again, then provide the contents of `/tmp/st` please?

> statusbar hangs up, too if:
> 
> 1. run i3 with i3status
> 2. change system time (one-two hours back)

That’s likely a separate problem. Can you file a separate issue for that please?

**Comment by szborows** [comment]
I confirm what @unera reports: it also happens on my machine, usually after suspending it. I'm using systemd. Sending SIGCONT to the process helps too.

Maybe the problem happens only when you go fullscreen and then suspend?

**Comment by szborows** [comment]
I was able to reproduce it with following scenario:

1. Switch to single display (laptop one)
2. Go fullscreen
3. Suspend
4. Un-suspend (restore?)
5. Before switching back to non-fullscreen mode switch desktop configuration (I'm using xrandr for this)
6. The process is stopped forever now. Even going back to non-fullscreen mode doesn't help.

**Comment by lasers** [comment]
Py3status have the same problem. See https://github.com/ultrabug/py3status/issues/941.

**Comment by steinex** [comment]
Same issue for me. I use `unredir-if-possible = true` in compton.conf. When compton unredirects e.g. due to opening a fullscreen video, i3bar sends SIGTSTP. However, sometimes it never sends the SIGCONT and i3status / py3status stop to update.

**Comment by steinex** [comment]
When it happens, running i3bar in verbose shows the following. Note: the bar is already visible again and the fullscreen video is stopped.

```
[i3bar/src/xcb.c:1907] Output xroot-0 inactive, skipping...
```

So it looks like i3bar is thinking that it's still invisible which it isn't.

**Comment by chbug** [comment]
Same thing here and on a workstation that doesn't go to sleep. I also suspect some interaction with fullscreen mode but haven't figured out the proper order of events that leads to this.

**Comment by nihr43** [comment]
My bar locks when I click or scroll on it.  I see nothing interesting in the debug logs, they just stop.  1.4-4
Should this conversation be over at vivien/i3blocks?

**Comment by jonpas** [comment]
Can also reproduce. Either automatically stops responding or I click on it (eg. calendar pop-up I have set up).

**Comment by Airblader** [comment]
If it hangs for the duration of a calendar being open or the like it might also be the process blocking the underlying status process.

**Comment by jonpas** [comment]
Calendar closes fine, it only recently started happening. I will investigate further (run without custom blocks etc) and see if I can get good reproduction steps.

**Comment by jonpas** [comment]
Looks like I jumped this issue a bit too soon. A block got updated and there was a bug where it requested superuser permissions, thus blocking all further execution if all blocks. Sorry for the inconvenience, feel free to delete my comments.

**Comment by wwalker** [comment]
I have the same problem.  Not sure what is triggering it.  I'm not suspending (it's a desk top and I never suspend it)  I've tried going full screen on windows on both monitors at the same time so it is hidden, and I've tried letting xscreensaver lock the screen.

I still don't know what is triggering it, but it is a SIGSTOP that is happening.

**Comment by wwalker** [comment]
As a temporary work around, you can make a script that does this:

```kill -SIGCONT $(ps -ef | grep i3_status_ba[r] | awk '{print $2}')```

replacing i3_status_ba[r] with the name of your status command.

**Comment by dfoxfranke** [comment]
I think I've figured this bug out.

https://github.com/i3/i3/blob/ca82f958125483a0e99c4ab82cda1b188754b32b/i3bar/src/xcb.c#L663-L674

Suppose that you have N>1 outputs with one bar on each output. Initially, suppose all outputs are active and all bars are visible. Then the following sequence of events occurs:

1. All bars become hidden.
2. All displays become inactive.
3. All bars become visible.
4. All displays become active.

After handling all the XCB events generated in step 1, the above code will count num_visible as 0 and send SIGSTOP to i3status. But when handling the events for step 3, it'll continue to count num_visible as 0 because it's skipping over all those inactive displays. So after step 4 when the displays become active again, the user will observe a stopped bar.

I have the following in my .config/i3/config, and I think it's plausibly triggering the above sequence of events and causing me to encounter this issue:

```
bindsym Pause      exec xset dpms force suspend; exec i3lock -e
```

Sometimes when unlocking my screen I type my password quickly enough that I've exited i3lock before my screens wake up, which would cause the bar visibility events of step 3 to trigger earlier than the display activity events of step 4.

**Comment by stapelberg** [comment]
> I think I've figured this bug out.

Thanks for looking into it! The detective work is much appreciated :)

> Suppose that you have N>=1 outputs with one bar on each output. Initially, suppose all outputs are active and all bars are visible. Then the following sequence of events occurs:
> 
> 1. All bars become hidden.
> 2. All outputs become inactive.
> 3. All bars become visible.
> 4. All outputs become active.
> 
> After handling all the XCB events generated in step 1, the above code will count num_visible as 0 and send SIGSTOP to i3status. But when handling the events for step 3, it'll continue to count num_visible as 0 because it's skipping over all those inactive outputs. So after step 4 when the outputs become active again, the user will observe a stopped bar.

Yep, this indeed sounds like a bug.

> I have the following in my .config/i3/config, and I think it's plausibly triggering the above sequence of events and causing me to encounter this issue:
> 
> ```
> bindsym Pause      exec xset dpms force suspend; exec i3lock -e
> ```

Outputs remain active even when you turn them off using DPMS. To be certain, I just verified this using:

```
% while :; do i3-msg -t get_outputs | jq .; sleep 1; done
```

**Comment by Alexander--** [comment]
Why does i3bar need to SIGSTOP i3status in the first place? Stopping a process, responsible for updating the clock — what could possibly go wrong?

**Comment by stapelberg** [comment]
i3bar suspends the status process when that process is invisible as an optimization to save battery life (no point in computing something that’s not visible).

**Comment by Alexander--** [comment]
> i3bar suspends the status process when that process is invisible as an optimization to save battery life

I have a single monitor and do not use suspend/resume, but often switch between VTs with different Xorg sessions. This bug triggers roughly once a week. Might happen more often, but I only notice it after clock displayed in the i3status drifts too much.

**Comment by stapelberg** [comment]
I suggest patching in https://github.com/i3/i3/pull/3761 until it’s merged and released, then.

**Comment by Alexander--** [comment]
Any news on when this will be fixed?

**Comment by stapelberg** [comment]
Once @dfoxfranke (or somebody else) moves pull request #3761 forward.

If you’re in a hurry, I recommend you build your own version of i3 with that pull request applied.

**Comment by Alexander--** [comment]
I don't like this feature and how it is implemented.

Signals aren't reliable form of communication. More importantly, the possibility of erroneously suspending a status bar really isn't worth the hassle of having that feature.

To makes matters worse, i3bar does not offer a way to opt-out of suspension. A process can't intercept SIGSTOP, so status bar implementations that expect to run at all times are screwed.

You are right, I am "in hurry", so I can send a pull request, that removes the code, sending SIGSTOP/SIGCONT. I don't want to take over #3761 and invest effort in fixing this feature only for it to get broken again later (and breaking it is apparently too easy).

**Comment by stapelberg** [comment]
> You are right, I am "in hurry", so I can send a pull request, that removes the code, sending SIGSTOP/SIGCONT.

I’m not convinced that removing the code is the best option.

If you don’t want to contribute a proper fix, that’s okay. But we definitely want to try fixing this issue before we remove the feature.

**Comment by Alexander--** [comment]
> I’m not convinced that removing the code is the best option.

Are you sure, that you aren't overestimating practical usefulness of SIGSTOP?

When people expect to preserve charge of notebook's battery, they suspend. This is more reliable that SIGSTOP and actually shuts down _all_ processes, that can cause power drain. 

Saving power when lid is closed (without suspending!) sounds like somewhat obscure use case, that comes in play once in a while, but does not make big difference. Yet you are triggering the behaviour in question for all devices — even ones without a battery. If you have concrete concerns about power use of i3status, wouldn't it be better to fix the specific problems with it? That would actually decrease power use for everyone and at all times — not just when lid is closed or monitor is shut down.

**Comment by stapelberg** [comment]
> Saving power when lid is closed (without suspending!) sounds like somewhat obscure use case

I agree, but that’s not the use-case for the SIGSTOP feature. Instead, the following two scenarios were the main motivation:

1. Running i3bar in hide mode (see https://i3wm.org/docs/userguide.html#_display_mode) defaults to not displaying the status bar, and only displays it when you press the modifier key.
2. While in fullscreen mode (e.g. reading a scientific paper), the bar is invisible, too.

**Comment by Airblader** [comment]
I don't have strong opinions, but we could also just stop the process and spawn a new one rather than suspend(?)

**Comment by Alexander--** [comment]
> I don't have strong opinions, but we could also just stop the process and spawn a new one rather than suspend(?)

This bug report isn't about suspension _method_.

The problem is: i3bar sometimes suspends i3status and never resumes it again. The method does not matter, because the status bar isn't being resumed at all.

**Comment by NeoRaider** [comment]
I've tried #3761 and it doesn't fix the issue - especially when using a compositing manager, i3status is pretty much stuck forever after changing display configuration, unless I SIGCONT it manually.

I would prefer a simple way to disable the SIGSTOP feature completely, as it's useless for me. Other applications I have always running use several magnitudes more CPU / battery in idle than i3bar/i3status ...

**Comment by rvalieris** [comment]
I ran into a similar issue, and I managed to turn off sig stop/count by setting the stop_signal/cont_signal options of the i3bar protocol: https://i3wm.org/docs/i3bar-protocol.html
I just set both signals to SIGUSR1 (10).
not sure how to set this on i3status however.. (I have my on custom i3bar script).

**Comment by indradhanush** [comment]
I was using `i3status` and encountered this problem. I recently moved to `i3status-rust` and I see the same problem. So maybe this is an independent issue of both `i3status` and `i3status-rust`. Also I'm on kernel `5.5.13` running Arch Linux. 

My `i3` version is `4.18`.

**Comment by Alexander--** [comment]
@indradhanush

This is upstream bug in i3. All i3 status bars, no matter the implementation language, are equally vulnerable to it. There is nothing that can be done about it by i3status or it's alternatives.

**Comment by orestisfl** [comment]
Please see if #3996 fixes your issue

**Comment by Alexander--** [comment]
@orestisfl

Thanks for fixing (?) this. In case this issue resurfaces again, is it possible to somehow make it more obvious, that status bar is hung and the status line isn't being updated?

Can i3 somehow fill the window of suspended status bar with solid color until it is repainted?

**Comment by orestisfl** [comment]
I don't see the value in going out of our way to implement that. I can put some debug log messages though.

**Comment by ammgws** [comment]
@orestisfl 
If I have understood correctly, #3996 does not solve the issue of the bar being paused when hidden, does it?

**Comment by orestisfl** [comment]
The bar is still paused, just continued more liberally
